src/workerd/api/tcp.h

```diff
@@ -0,0 +1,130 @@
+
+  // AsyncOutputStream methods:
+  kj::Promise<void> write(const void* buffer, size_t size) override;
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override;
+
+  kj::Promise<void> whenWriteDisconnected() override;
+
+private:
+  // Either calls `then` on the `inner` promise with the `f` callback, or calls `f` on an already
+  // stored stream (if the `inner` promise completed).
+  template <typename T>
+  kj::Promise<T> thenOrRunNow(std::function<kj::Promise<T> (kj::Own<kj::AsyncIoStream> *)> f) {
+    KJ_IF_MAYBE(e, error) {
+      kj::throwRecoverableException(kj::mv(*e));
+    }
+
+    KJ_IF_MAYBE(s, completedInner) {
+      return f(s);
+    } else {
+      return inner.then([this, f](kj::Own<kj::AsyncIoStream> stream) -> kj::Promise<T> {
+        completedInner = kj::mv(stream);
+        return f(&KJ_ASSERT_NONNULL(completedInner));
+      });
+    }
+  }
+
+  // Stored io stream once the `inner` promise completes.
+  kj::Maybe<kj::Own<kj::AsyncIoStream>> completedInner;
+  // Stored error if any of the operations throw.
+  kj::Maybe<kj::Exception> error;
+  // A promise holding a stream that will be available at some point in the future. Typically once
+  // a connection is established.
+  kj::Promise<kj::Own<kj::AsyncIoStream>> inner;
+};
+
+struct SocketOptions: public jsg::Object {
+  // TODO(later): TCP socket options will live here.
+  SocketOptions() {};
+public:
+  JSG_RESOURCE_TYPE(SocketOptions, CompatibilityFlags::Reader flags) { }
+};
+
+class Socket: public jsg::Object {
+public:
+  Socket(kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise);
+
+  kj::Promise<kj::HttpClient::ConnectResponse> connectionPromise;
+
+  jsg::Ref<ReadableStream> getReadable() { return KJ_ASSERT_NONNULL(readable).addRef(); }
+  jsg::Ref<WritableStream> getWritable() { return KJ_ASSERT_NONNULL(writable).addRef(); }
+  jsg::Promise<void> getClosed() {
+    // TODO: Right now this promise won't complete when the remote end closes the socket.
+    //       Do we need to wrap ReadableStream/WritableStream to make this work or is there a
+    //       better way?
+    auto& context = IoContext::current();
+    return context.awaitIo(kj::mv(closeFulfiller->promise));
+  }
+
+  jsg::Promise<void> close(jsg::Lock& js);
+  // Closes the socket connection.
+
+  JSG_RESOURCE_TYPE(Socket, CompatibilityFlags::Reader flags) {
+    if (flags.getJsgPropertyOnPrototypeTemplate()) {
+      JSG_READONLY_PROTOTYPE_PROPERTY(readable, getReadable);
+      JSG_READONLY_PROTOTYPE_PROPERTY(writable, getWritable);
+      JSG_READONLY_PROTOTYPE_PROPERTY(closed, getClosed);
+    } else {
+      JSG_READONLY_INSTANCE_PROPERTY(readable, getReadable);
+      JSG_READONLY_INSTANCE_PROPERTY(writable, getWritable);
+      JSG_READONLY_INSTANCE_PROPERTY(closed, getClosed);
+    }
+    JSG_METHOD(close);
+  }
+
+private:
+  kj::Maybe<jsg::Ref<ReadableStream>> readable;
+  kj::Maybe<jsg::Ref<WritableStream>> writable;
```

Comment:
This is likely fine for now, but why are these in maybes? If it is because the streams might not yet be initialized when the object is created, then if the `Socket` gets passed out to JS while those are still initializing, the `KJ_ASSERT_NONNULL(...)`s in the `getReadable`/`getWritable` will be problematic. What I would expect is that these are just allocated immediately when the `Socket` is created and that the underlying `ReadableStreamSource` and `WritableStreamSink` impls would handle things appropriately internally.

Reply:
It's because I had to initialise them in the constructor here: https://github.com/cloudflare/workerd/pull/162/files#diff-7aaaf2cec480ca0e3779eeec2a72738437ca401a81331b2f3706d16e598e7ba7R103. Maybe this is my C++ inexperience showing, but the compiler didn't like that I wasn't initialising it in the constructor's initialiser list so I turned it into a Maybe to satisfy it. Is there a better way? hm, maybe I can create a separate function and call that in the initialiser list... but then this function would need to return a struct containing the two streams. Would that be preferable? or is there a better approach I am missing?

Reply:
Spoke with harris and he suggested using a dependent constructor. TIL this is possible in C++. Anyway, updated with that and now no more `Maybe` :)